### PR TITLE
test(plugin): proptests for implicit_prices, sell_gains, coherent_cost, check_average_cost (closes #1004)

### LINKED ISSUE #1004: Phase 4: property tests for numeric plugins
## Background

Phase 4 of the plugin-testing-quality plan (see #992). Plugins that compute new numbers from existing ones have algebraic invariants that example tests miss because the input space is too large.

`implicit_prices` (#992): the property "emitted price round-trips through PriceDatabase to the original per-unit value" would have crashed CI on the first generated `@@` input. A property test would have caught the bug 100% of the time.

## What this issue tracks

Add at least one `proptest` case to each numeric plugin covering an algebraic invariant.

## Suggested properties

| Plugin | Invariant to pin |
|---|---|
| `implicit_prices` | `PriceDatabase::get_price(emitted) == original per-unit price` |
| `unrealized` | `Sum(realized) + Sum(unrealized) = Sum(total)` at any cutoff date |
| `sell_gains` | Gains posting balances the transaction |
| `coherent_cost` | Same lot at same date has same cost basis |
| `check_average_cost` | Average of N positions equals weighted mean of inputs |

## Implementation notes

- 256 proptest cases per CI run (default) is enough.
- Use `proptest::collection::vec` to generate small lists of postings.
- Filter out degenerate inputs (zero amounts, NaN-like Decimals) early so the test focuses on the property, not edge-case handling.

## Acceptance criteria

- [ ] Each numeric plugin has at least one proptest case in `crates/rustledger-plugin/tests/native_plugins_test.rs`
- [ ] Properties are documented inline (what invariant, why it should hold)
- [ ] CI runs them on every PR (proptest is a test-time crate, no build config changes needed)

## Why this matters even after #992 is fixed

The shared helper from #997 means the bug class is closed for `implicit_prices` specifically. But the next numeric plugin will reintroduce the risk unless property tests force the author to think about invariants from day one. Mutation testing (#1003) catches some of this; property tests are complementary — they explore the space rather than just resisting mutation.

## Related

- #992 — the original bug
- #1001 — Phase 5 policy (which now requires proptest for new numeric plugins)
- #1003 — Phase 3 mutation testing (complementary)

### PR DESCRIPTION
**Pull request by robcohen**
## Summary

Phase 4 of the plugin-testing-quality plan (#992). Adds property tests covering algebraic invariants for the four numeric plugins listed in #1004 that lacked any proptest coverage. The fifth plugin in the issue (`unrealized`) already had a proptest from #1007.

## Properties pinned

| Plugin | Invariant | Why it matters |
|---|---|---|
| `implicit_prices` | `N C @@ (N*P) Q` → emitted Price = `C @ P Q` exactly | EXACT regression test for #992 (off-by-factor-of-N bug) |
| `sell_gains` | warning iff `(sale - cost) * |units| ≠ 0 AND no Income posting` | Sweeps gain, loss, and zero-gain cases independently |
| `coherent_cost` | error count = `|with_cost ∩ price_only|` | Catches classification drift in the cost/price bookkeeping |
| `check_average_cost` | selling at computed weighted mean → 0 warnings | Rounding/off-by-one in running totals would break this |

For `check_average_cost` the test computes the expected mean using `rust_decimal` (same crate as the plugin) so any precision loss is identical on both sides — a real bug, not a precision artifact, would be required to fail.

## Test count

- `native_plugins_test`: 167 → **171** (4 new proptests, each 64 cases per run)
- All proptests use `proptest::collection::vec` for posting-list inputs and `Decimal::new(cents, 2)` for prices to keep arithmetic well inside `rust_decimal` precision.

## Test plan

- [x] `cargo test -p rustledger-plugin` — all 171 tests pass (proptests + examples)
- [x] `cargo clippy -p rustledger-plugin --all-features --all-targets -- -D warnings` — clean (after fixing two `uninlined_format_args` lints)
- [x] `cargo fmt -p rustledger-plugin -- --check` — clean

Closes #1004

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/rustledger-plugin/tests/native_plugins_test.rs [hunk 1]
```diff
@@ -2654,6 +2654,71 @@ fn test_implicit_prices_emitted_excludes_input_price_directives() {
...
+        let plugin = ImplicitPricesPlugin;
+        let input = make_input(vec![
+            make_open("2024-01-01", "Assets:Brokerage"),
+            make_open("2024-01-01", "Assets:Cash"),
+            make_transaction_with_cost_and_price_total(
+                "2024-01-15",
+                "Buy at total price",
+                "Assets:Brokerage",
+                (&units.to_string(), "HOOL"),
+                ("0", "USD"), // cost stub — per-unit annotation should win
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-plugin/tests/native_plugins_test.rs]: Good catch — the test only constructed `@@` and the inline cost-stub comment was wrong about "per-unit annotation". Fixed in [`ba0474a4`](https://github.com/rustledger/rustledger/pull/1042/commits/ba0474a4): restructured to use `make_txn_with_price_annotation` with a generated `is_total` bool, so both forms are now actually exercised in the same proptest (renamed `prop_implicit_prices_emits_per_unit_for_both_annotation_forms`). The `@@` branch remains the EXACT #992 regression; the `@` branch pins the trivial-but-mutable identity path.

File: crates/rustledger-plugin/tests/native_plugins_test.rs [hunk 2]
```diff
@@ -6624,6 +6915,97 @@ fn test_check_average_cost_respects_custom_tolerance() {
...
+            1..=5,
+        ),
+    ) {
+        use rust_decimal::Decimal;
+
+        // Compute the expected weighted mean. Sum cost as
+        // Σ(units * price) in cents, then divide by total units, then
+        // divide by 100 to convert cents to dollars. Using the same
+        // operation order as the plugin (`total_cost / total_units`)
+        // ensures that any precision loss is identical on both sides.
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger-plugin/tests/native_plugins_test.rs]: Right — the comment described an arithmetic flow the code does not perform. `Decimal::new(price_cents, 2)` already encodes the price in dollars (scale 2 ⇒ value = cents/100), so there is no separate `/ 100` step. Rewritten in [`ba0474a4`](https://github.com/rustledger/rustledger/pull/1042/commits/ba0474a4) to describe what the code actually does.
